File: backend/app/schemas/trace.py
from datetime import datetime

from pydantic import BaseModel, Field


class TraceIn(BaseModel):
    model: str
    provider: str
    prompt: str
    completion: str | None = None
    prompt_tokens: int = 0
    completion_tokens: int = 0
    latency_ms: float
    cost: float = 0.0
    status: str = "success"
    error_message: str | None = None
    tags: dict = Field(default_factory=dict)
    # Set by the SDK so a retried batch is a no-op instead of a duplicate.
    # Optional because manual API callers (e.g. via /docs) won't set it.
    client_trace_id: str | None = None


class TraceBatchIn(BaseModel):
    traces: list[TraceIn] = Field(min_length=1, max_length=500)


class TraceOut(BaseModel):
    id: str
    project_id: str
    client_trace_id: str | None
    model: str
    provider: str
    prompt: str
    completion: str | None
    prompt_tokens: int
    completion_tokens: int
    latency_ms: float
    cost: float
    status: str
    error_message: str | None
    tags: dict
    created_at: datetime

    model_config = {"from_attributes": True}


class TraceListOut(BaseModel):
    items: list[TraceOut]
    total: int
    limit: int
    offset: int

File: dashboard/src/types/trace.ts
// Mirrors app/schemas/trace.py in the backend.

/**
 * A single trace record returned by both the list and detail endpoints.
 * Field names match the backend TraceOut Pydantic schema EXACTLY.
 */
export interface TraceOut {
  id: string;
  project_id: string;
  client_trace_id: string | null; // SDK-generated idempotency ID, may be null for manual traces
  model: string;                  // e.g. "gpt-4o", "claude-3-5-sonnet-20241022"
  provider: string;               // e.g. "openai", "anthropic"
  prompt: string;
  completion: string | null;      // null if the call errored before a response
  prompt_tokens: number;
  completion_tokens: number;
  latency_ms: number;             // milliseconds as a float
  cost: number;                   // USD as a float, e.g. 0.00032
  status: string;                 // "success" | "error"
  error_message: string | null;   // only set when status = "error"
  tags: Record<string, unknown>;  // arbitrary JSON object — e.g. { "env": "prod", "version": "2" }
  created_at: string;             // ISO datetime string
}

/**
 * Returned by GET /api/v1/projects/:id/traces
 * A paginated wrapper around a list of traces.
 */
export interface TraceListOut {
  items: TraceOut[];
  total: number;  // total matching records in the DB (across all pages)
  limit: number;  // how many were requested (default 50)
  offset: number; // how many were skipped (default 0)
}

/**
 * Parameters for the list traces endpoint.
 * All fields are optional filters.
 */
export interface TraceFilters {
  model?: string;
  status?: string;
  limit?: number;
  offset?: number;
}
